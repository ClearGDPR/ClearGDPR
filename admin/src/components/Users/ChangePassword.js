import React from 'react';
import PropTypes from 'prop-types';
import { Form } from 'react-form';
import TextInput from '../core/cards/dashboard/TextInput';
import Loader from '../core/cards/dashboard/Loader';

export class ChangePassword extends React.Component {
  static propTypes = {
    isLoading: PropTypes.bool,
    onSubmit: PropTypes.func,
    touched: PropTypes.object,
    errors: PropTypes.object
  };

  validatePassword(password) {
    // TODO: todo add proper validation
    if (!password) {
      return {
        error: 'Field required'
      };
    }
    if (password.length < 8) {
      return {
        error: 'Password must have min. 8 characters'
      };
    }
    return {
      success: null
    };
  }

  renderForm() {
    return (
      <React.Fragment>
        <TextInput
          label="New password"
          placeholder="*********"
          error={
            this.props.touched &&
            this.props.touched['newPassword'] &&
            this.props.errors &&
            this.props.errors['newPassword']
          }
          field="newPassword"
          validate={this.validatePassword}
        />
        <TextInput
          label="Repeat new password"
          placeholder="*********"
          error={
            this.props.touched &&
            this.props.touched['newPasswordRepeat'] &&
            this.props.errors &&
            this.props.errors['newPasswordRepeat']
          }
          field="newPasswordRepeat"
          validate={this.validatePassword}
        />
        <div>
          <input type="submit" className="btn" value="Save" />
        </div>
      </React.Fragment>
    );
  }

  render() {
    return (
      <form onSubmit={this.props.onSubmit}>
        {this.props.isLoading ? <Loader /> : this.renderForm()}
      </form>
    );
  }
}

const changePasswordForm = props => (
  <Form onSubmit={submittedValues => props.onSubmit && props.onSubmit(submittedValues)}>
    {formApi => (
      <ChangePassword
        onSubmit={formApi.submitForm}
        errors={formApi.errors}
        touched={formApi.touched}
        {...props}
      />
    )}
  </Form>
);

changePasswordForm.propTypes = {
  onSubmit: PropTypes.func
};

export default changePasswordForm;
