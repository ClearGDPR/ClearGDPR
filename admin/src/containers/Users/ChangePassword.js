import React from 'react';
import PropTypes from 'prop-types';

import ChangePassword from 'components/Users/ChangePassword';
import config from 'config';
import { PanelConsumer } from 'containers/MainLayout/PanelContext';
import internalFetch from 'helpers/internal-fetch';
import { toast } from 'react-toastify';

export class ChangePasswordContainer extends React.Component {
  static propTypes = {
    userId: PropTypes.number.isRequired,
    closePanel: PropTypes.func
  };

  state = {
    isLoading: false,
    errors: {}
  };

  async changePassword(userId, password) {
    await internalFetch(`${config.API_URL}/api/management/users/${userId}/update-password`, {
      method: 'POST',
      body: JSON.stringify({
        password
      })
    }).catch(err => {
      this.setState({
        errors: {
          '': err.message
        }
      });
      return Promise.reject(err);
    });
  }

  startLoading() {
    this.setState({
      isLoading: true
    });
  }

  validatePassword(password) {
    // TODO: todo add proper validation
    if (!password) {
      return 'Field required';
    }
    if (password.length < 8) {
      return 'Password must have min. 8 characters';
    }
    return null;
  }

  stopLoading = () => {
    this.setState({
      isLoading: false
    });
  };

  onSubmit = data => {
    const { newPassword, newPasswordRepeat } = data;

    if (newPassword !== newPasswordRepeat) {
      this.setState({
        errors: {
          newPasswordRepeat: 'Passwords do not match!'
        }
      });
      return;
    }

    this.startLoading();

    return this.changePassword(this.props.userId, newPassword)
      .then(this.stopLoading)
      .then(res => {
        toast.success('User successfully registered');
        return res;
      })
      .then(() => this.props.closePanel && this.props.closePanel())
      .catch(this.stopLoading);
  };

  render() {
    return (
      <ChangePassword
        errors={this.state.errors}
        validatePassword={this.validatePassword}
        onSubmit={this.onSubmit}
        isLoading={this.state.isLoading}
      />
    );
  }
}

export default props => (
  <PanelConsumer>
    {({ closePanel }) => <ChangePasswordContainer {...props} closePanel={closePanel} />}
  </PanelConsumer>
);
