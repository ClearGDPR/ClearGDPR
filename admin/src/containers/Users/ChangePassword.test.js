import React from 'react';
import { shallow } from 'enzyme';

import { ChangePasswordContainer as ChangePassword } from './ChangePassword';

const setupShallow = propOverrides => {
  const props = Object.assign({ userId: 1 }, propOverrides);
  const component = shallow(<ChangePassword {...props} />);

  return { props, component };
};

describe('(Component) Change Password', () => {
  it('should render correctly when user id provided props provided', async () => {
    const { component } = setupShallow();
    expect(component).toMatchSnapshot();
  });

  describe('Validate password', () => {
    it('should return error when password is null or empty', async () => {
      const { component } = setupShallow();
      let result = component.instance().validatePassword('');
      expect(result).toEqual(
        expect.objectContaining({
          error: 'Field required'
        })
      );
      result = component.instance().validatePassword();
      expect(result).toEqual(
        expect.objectContaining({
          error: 'Field required'
        })
      );
    });

    it('should return error when password is too short', async () => {
      const { component } = setupShallow();
      let result = component.instance().validatePassword('asc');
      expect(result).toEqual(
        expect.objectContaining({
          error: 'Password must have min. 8 characters'
        })
      );
      result = component.instance().validatePassword('asffas');
      expect(result).toEqual(
        expect.objectContaining({
          error: 'Password must have min. 8 characters'
        })
      );
    });
  });
});
