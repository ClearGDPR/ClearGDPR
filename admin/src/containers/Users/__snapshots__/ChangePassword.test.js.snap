// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Component) Change Password should render correctly when user id provided props provided 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ChangePasswordContainer
    userId={1}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "errors": Object {},
      "isLoading": false,
      "onSubmit": [Function],
      "validatePassword": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "errors": Object {},
        "isLoading": false,
        "onSubmit": [Function],
        "validatePassword": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`(Container) Change Password should render correctly when user id provided props provided 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ChangePasswordContainer
    userId={1}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "errors": Object {},
      "isLoading": false,
      "onSubmit": [Function],
      "validatePassword": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "errors": Object {},
        "isLoading": false,
        "onSubmit": [Function],
        "validatePassword": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
